### PR TITLE
pseudoinverse pull request #2

### PR DESCRIPTION
**Pull request by anish-lakkapragada**
created a second psuedoinverse PR because the first one was too messy. 

defn psuedoinverse function in Nx.LinAlg just as np.linalg.svd in numpy. Uses the SVD to compute the psuedoinverse.

Works only for wide matrices with backends that have SVD implemented for wide matrices

### CODE REVIEW

File: nx/lib/nx/lin_alg.ex [hunk 1]
```diff
@@ -949,6 +949,46 @@ defmodule Nx.LinAlg do
...
+
+      iex> pinv = Nx.LinAlg.pinv(Nx.tensor([[1, 1], [3, 4]]))
+      iex> pinv
+      #Nx.Tensor<
+        f32[2][2]
+        [
+          [4.0, -1.0000001192092896],
+          [-3.0, 1.0000001192092896]
+        ]
+      >
```
Comment (seanmor5) [inline nx/lib/nx/lin_alg.ex:968]:
```suggestion
      >

```

File: nx/lib/nx/lin_alg.ex [hunk 2]
```diff
@@ -949,6 +949,46 @@ defmodule Nx.LinAlg do
...
+      iex> pinv = Nx.LinAlg.pinv(Nx.tensor([[1, 3], [5, 4], [4, 9]]))
+      iex> pinv
```
Comment (seanmor5) [inline nx/lib/nx/lin_alg.ex:970]:
```suggestion
      iex> Nx.LinAlg.pinv(Nx.tensor([[1, 3], [5, 4], [4, 9]]))
```

File: nx/lib/nx/lin_alg.ex [hunk 3]
```diff
@@ -949,6 +949,46 @@ defmodule Nx.LinAlg do
...
+  @doc """
+  Calculates the Moore-Penrose inverse, or the pseudoinverse, of a matrix.
+
+  ## Options
+    * `:eps` - Rounding error threshold used to assume values as 0.
+
+  ## Examples
+
+      iex> pinv = Nx.LinAlg.pinv(Nx.tensor([[1, 1], [3, 4]]))
+      iex> pinv
```
Comment (seanmor5) [inline nx/lib/nx/lin_alg.ex:961]:
```suggestion
      iex> Nx.LinAlg.pinv(Nx.tensor([[1, 1], [3, 4]]))
```

File: nx/lib/nx/lin_alg.ex [hunk 4]
```diff
@@ -949,6 +949,46 @@ defmodule Nx.LinAlg do
...
+      >
+  """
+  defn pinv(tensor, opts \\ []) do
+
+    opts = keyword!(opts, eps: @default_eps)
+
+    {u, s, vt} = Nx.LinAlg.svd(tensor)
+
+    s_shape = {elem(Nx.shape(s), 0), elem(Nx.shape(u), 1)}
+    adjusted_s = Nx.map(s, fn x -> Nx.select(Nx.abs(x) < opts[:eps], 0, 1 / x) end)
```
Comment (seanmor5) [inline nx/lib/nx/lin_alg.ex:986]:
```suggestion
    adjusted_s = Nx.select(Nx.abs(x) < opts[:eps], 0, 1 / x)
```

You should basically never use Nx.map, Nx.select works element-wise

File: nx/lib/nx/lin_alg.ex [hunk 5]
```diff
@@ -949,6 +949,46 @@ defmodule Nx.LinAlg do
...
+    s_matrix = Nx.broadcast(0, s_shape) |> Nx.put_diagonal(adjusted_s)
+
+    Nx.transpose(vt) |> Nx.dot(s_matrix) |> Nx.dot(Nx.transpose(u))
```
Comment (polvalente) [inline nx/lib/nx/lin_alg.ex:989]:
```suggestion
    adjoint(vt) |> Nx.dot(s_matrix) |> Nx.dot(adjoint(u))
```

The transpose only works for real tensors. A separate test for complex should be added as well (though I don't think we need to clutter examples with this)

File: nx/lib/nx/lin_alg.ex (hunk 3 above)
Comment (anish-lakkapragada) [reply nx/lib/nx/lin_alg.ex]: should be fixed in commit 14811dbec97df7007212336b3131042afa8079cc. deleted the extra line

File: nx/lib/nx/lin_alg.ex (hunk 4 above)
Comment (anish-lakkapragada) [reply nx/lib/nx/lin_alg.ex]: should be fixed in 14811dbec97df7007212336b3131042afa8079cc. Used Nx.select and no Nx.map

File: nx/lib/nx/lin_alg.ex (hunk 3 above)
Comment (anish-lakkapragada) [reply nx/lib/nx/lin_alg.ex]: update: 4a8f75a0e0f9c22f7c87805b12ccd809332e1e9a

File: nx/lib/nx/lin_alg.ex (hunk 4 above)
Comment (anish-lakkapragada) [reply nx/lib/nx/lin_alg.ex]: update on latest commit: 4a8f75a0e0f9c22f7c87805b12ccd809332e1e9a

File: nx/lib/nx/lin_alg.ex (hunk 5 above)
Comment (anish-lakkapragada) [reply nx/lib/nx/lin_alg.ex]: I've added `adjoint(vt)` but still need to test for complex matrices.

Comment (anish-lakkapragada) [reply nx/lib/nx/lin_alg.ex]: It appears as though the `svd` at least for the default backend is not implemented for complex inputs.

File: nx/lib/nx/lin_alg.ex [hunk 6]
```diff
@@ -949,6 +949,44 @@ defmodule Nx.LinAlg do
...
+      iex> Nx.LinAlg.pinv(Nx.tensor([[1, 3], [5, 4], [4, 9]]))
+      #Nx.Tensor<
+        f32[2][3]
+        [
+          [-0.07312049716711044, 0.3027806580066681, -0.11019568890333176],
+          [0.06900102645158768, -0.1307930052280426, 0.1462409943342209]
+        ]
+      >
+  """
+  defn pinv(tensor, opts \\ []) do
```
Comment (josevalim) [inline nx/lib/nx/lin_alg.ex:978]: Thank you. Let’s rename the function to pseudo_inverse?

We also need to make sure it works batched. If you pass a higher than rank-2 tensor, we should assume that each additional dimension works as a batch. :)

Comment (polvalente) [reply nx/lib/nx/lin_alg.ex:978]: The only issue I have with `pseudo_inverse` is that for normal inverse we use `Nx.LinAlg.invert`. Should we change that to `inverse`?

File: nx/lib/nx/lin_alg.ex (hunk 5 above)
Comment (polvalente) [reply nx/lib/nx/lin_alg.ex]: Don't worry about this :)
I'm working on #1015 which reimplements SVD as whole, so I can deal with this afterwards

Comment (anish-lakkapragada) [reply nx/lib/nx/lin_alg.ex]: Ah, I see. I will wait for any other feedback

File: nx/lib/nx/lin_alg.ex (hunk 6 above)
Comment (anish-lakkapragada) [reply nx/lib/nx/lin_alg.ex:978]: for more context, `np.linalg.pinv` is how it is done in numpy

File: nx/lib/nx/lin_alg.ex [hunk 7]
```diff
@@ -949,6 +949,44 @@ defmodule Nx.LinAlg do
...
+      >
+
+      iex> Nx.LinAlg.pinv(Nx.tensor([[1, 3], [5, 4], [4, 9]]))
+      #Nx.Tensor<
+        f32[2][3]
+        [
+          [-0.07312049716711044, 0.3027806580066681, -0.11019568890333176],
+          [0.06900102645158768, -0.1307930052280426, 0.1462409943342209]
+        ]
+      >
```
Comment (polvalente) [inline nx/lib/nx/lin_alg.ex:976]:
```suggestion
      iex> Nx.LinAlg.pinv(Nx.tensor(([[0.5, 0], [0, 1], [0.5, 0]]))
      #Nx.Tensor<
        f32[2][3]
        [
          [1.0, 0.0, 1.0],
          [0.0, 1.0, 0.0]
        ]
      >
```

There might be a rounding error or 2, but in general this example might look cleaner for a wide matrix

File: nx/lib/nx/lin_alg.ex (hunk 6 above)
Comment (josevalim) [reply nx/lib/nx/lin_alg.ex:978]: @polvalente should we call it `pseudo_invert` then? or `pinv`? What do you decide?

Comment (polvalente) [reply nx/lib/nx/lin_alg.ex:978]: pinv is a really usual name for this function

All of the following langs/libs use it: NumPy/Jax, PyTorch, Julia, MATLAB/Octave

So I think it's ok to go with pinv :)

File: nx/lib/nx/lin_alg.ex [hunk 8]
```diff
@@ -949,6 +949,44 @@ defmodule Nx.LinAlg do
...
+      #Nx.Tensor<
+        f32[2][3]
+        [
+          [1.0, 0.0, 1.0],
+          [0.0, 1.0, 0.0]
+        ]
+      >
+  """
+  defn pinv(tensor, opts \\ []) do
+
```
Comment (josevalim) [inline nx/lib/nx/lin_alg.ex:979]:
```suggestion
```

**Review by josevalim — APPROVED** [review]

File: nx/lib/nx/lin_alg.ex [hunk 9]
```diff
@@ -949,6 +949,43 @@ defmodule Nx.LinAlg do
...
+          [4.0, -1.0000001192092896],
+          [-3.0, 1.0000001192092896]
+        ]
+      >
+
+      iex> Nx.LinAlg.pinv(Nx.tensor([[0.5, 0], [0, 1], [0.5, 0]]))
+      #Nx.Tensor<
+        f32[2][3]
+        [
+          [1.0, 0.0, 1.0],
```
Comment (polvalente) [inline nx/lib/nx/lin_alg.ex:973]:
```suggestion
          [0.9999999403953552, 0.0, 0.9999999403953552],
```

File: nx/lib/nx/lin_alg.ex [hunk 10]
```diff
@@ -949,6 +949,43 @@ defmodule Nx.LinAlg do
...
+
+  ## Examples
+
+      iex> pinv = Nx.LinAlg.pinv(Nx.tensor([[1, 1], [3, 4]]))
+      #Nx.Tensor<
+        f32[2][2]
+        [
+          [4.0, -1.0000001192092896],
+          [-3.0, 1.0000001192092896]
+        ]
```
Comment (polvalente) [inline nx/lib/nx/lin_alg.ex:966]:
```suggestion
        [
          [3.9924840927124023, -1.0052788257598877],
          [-3.005120038986206, 1.0071183443069458]
        ]
```

File: nx/lib/nx/lin_alg.ex [hunk 11]
```diff
@@ -949,6 +949,43 @@ defmodule Nx.LinAlg do
     )
   end
 
+  @doc """
+  Calculates the Moore-Penrose inverse, or the pseudoinverse, of a matrix.
+
+  ## Options
+    * `:eps` - Rounding error threshold used to assume values as 0.
```
Comment (polvalente) [inline nx/lib/nx/lin_alg.ex:956]:
```suggestion
    * `:eps` - Rounding error threshold used to assume values as 0. Defaults to `1.0e-10`
```

File: nx/lib/nx/lin_alg.ex [hunk 12]
```diff
@@ -949,6 +949,43 @@ defmodule Nx.LinAlg do
...
+      #Nx.Tensor<
+        f32[2][3]
+        [
+          [1.0, 0.0, 1.0],
+          [0.0, 1.0, 0.0]
+        ]
+      >
+  """
+  defn pinv(tensor, opts \\ []) do
+    opts = keyword!(opts, eps: @default_eps)
```
Comment (polvalente) [inline nx/lib/nx/lin_alg.ex:979]:
```suggestion
    opts = keyword!(opts, eps: 1.0e-10)
```

**Review by polvalente — APPROVED** [review]

File: nx/lib/nx/lin_alg.ex [hunk 13]
```diff
@@ -949,6 +949,43 @@ defmodule Nx.LinAlg do
     )
   end
 
+  @doc """
+  Calculates the Moore-Penrose inverse, or the pseudoinverse, of a matrix.
+
+  ## Options
```
Comment (polvalente) [inline nx/lib/nx/lin_alg.ex:955]:
```suggestion
  ## Options
  
```

File: nx/lib/nx/lin_alg.ex [hunk 14]
```diff
@@ -949,6 +949,43 @@ defmodule Nx.LinAlg do
...
+    * `:eps` - Rounding error threshold used to assume values as 0. Defaults to `1.0e-10`
+
+  ## Examples
+
+      iex> pinv = Nx.LinAlg.pinv(Nx.tensor([[1, 1], [3, 4]]))
```
Comment (polvalente) [inline nx/lib/nx/lin_alg.ex:960]:
```suggestion
      iex> Nx.LinAlg.pinv(Nx.tensor([[1, 1], [3, 4]]))
```

**Pull request merged**